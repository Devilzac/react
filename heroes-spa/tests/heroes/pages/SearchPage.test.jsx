const { render, screen, getByRole, fireEvent } = require("@testing-library/react")
const { MemoryRouter,useNavigate } = require("react-router-dom")
const { SearchPage } = require("../../../src/heroes/pages/SearchPage")
const { default: queryString } = require("query-string")


const mockedUseNavigate = jest.fn();
const mockedSubmit = jest.fn();


jest.mock('react-router-dom', ()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate: ()=>mockedUseNavigate,
}));

describe('pruebas en <SearchPage />', () => { 
    
    beforeEach(()=> jest.clearAllMocks());

    test('Debe de mostrarse correctamente con valores por defecto  ', () => { 

        const {container} = render(
            <MemoryRouter>
                <SearchPage/>
            </MemoryRouter>
        )
        
        expect(container).toMatchSnapshot()
       
    })
    
    test('Debe de mostrar a Batman y el input con el valor del query string  ', () => { 
        const {container} = render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage/>
            </MemoryRouter>
        )
        const result = screen.getByRole('textbox').value
        const img = screen.getByRole('img').src
        const dev = screen.getByLabelText('alert-danger').style.display
        
        expect(result).toBe('batman')
        expect(img).toContain('/src/assets/heroes/dc-batman.jpg')
        expect(dev).toBe('none')
       
    })

    test('debe de mostrar un error si no se encuentra el heroe', () => { 

        const {container} = render(
            <MemoryRouter initialEntries={['/search?q=asd']}>
                <SearchPage/>
            </MemoryRouter>
        )
        const dev = screen.getByLabelText('alert-danger').style.display
        expect(dev).toBe('')

     })

     test('debe de llamar el navigate a la pantalla nueva', () => { 
                
        const {container} = render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage/>
            </MemoryRouter>
        )
        
        const input = screen.getByRole('textbox');
        const form = screen.getByTestId('searchHero');
        const value = 'batman'

        fireEvent.change(input, {target:{
            name: 'searchText',
            value: value
        }});


        fireEvent.submit(form);
        expect(input.value).toBe(value);
        expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${value}`);

        
      })

 })