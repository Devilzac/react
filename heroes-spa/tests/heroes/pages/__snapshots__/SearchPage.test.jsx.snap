// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <SearchPage /> Debe de mostrarse correctamente con valores por defecto   1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        data-testid="searchHero"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a Hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        search a hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b>
           
           
        </b>
      </div>
    </div>
  </div>
</div>
`;
